### ISSUE TITLE
"Review Github pull request" feature

### ISSUE THREAD

**Issue by stevage** [divmain/GitSavvy#308]
There doesn't seem to be a good workflow in GitSavvy for evaluating pull requests, so I end up dropping into a command line for this. I imagine it working like this:

1. Cmd-Shift-P, "Checkout Github pull request"
2. Paste pull request URL: https://github.com/myorg/myproj/pull/666
3. GitSavvy magically figures out that this corresponds to branch "mycoolfeature" at https://github.com/newcontributor/myproj
4. GitSavvy executes these commands:

```
git checkout -b newcontributor-mycoolfeature
git pull https://github.com/newcontributor/myproj mycoolfeature
```

**Comment by divmain** [comment]
Generally, I've used `github: add fork as remote`, fetched, and checked out the branch.  But I don't mind streamlining this.

**Comment by divmain** [comment]
One problem with the provided commands in Step 4: If there is a diverged commit history, this will generate a merge commit, which you may not want.  Alternative would be:

- ensure the index is clean
- query for open pull requests
- upon selection, set a remote for the PR's source
- fetch new remote
- detach head at associated commit

This would be a clean flow for reviewing changes in a PR, without much potential for dirtying the local environment.

**Comment by stevage** [comment]
I probably should have mentioned the source of the two git commands at the end of the issue. That's actually the text that Github generates when you click "view command line instructions". 

Github's presumption is that you will review this branch, and then if it's good, switch back to master, merge the branch in (`git merge --no-ff branchname`), and push that. So generating extra merge commits on this branch is kind of irrelevant.

I like your proposed workflow (especially "query for open pull requests"), but personally I really don't like adding extra remotes. They seem to get in the way a lot - especially because they often have all the same branches, meaning you get n(branches) x n(remotes).

Functionally, there's very little difference between:

- adding a remote and checking out a detached head; and
- creating a temporary branch and pulling an arbitrary URL to it

except that in the latter case, you haven't added an extra remote, which I consider a bonus. You could even name the temporary branch something like "review-nnn-xxx" where nnn is the pull request number, and xxx is the branch name.

**Comment by divmain** [comment]
However, if the the PR's `HEAD` has a merge conflict with the latest master on _your_ machine, the pull will not be successful.  Fortunately, I've found a neat little workaround.
```
git fetch https://github.com/user/remote.git PR-SHA
```
No need to add a remote.  This will fetch that commit and the smallest possible set of ancestor commits, at which point it can be checked out as detached HEAD (or a branch can be created).  And any complications with merging can be avoided.

**Comment by divmain** [comment]
@stevage, give this a try when you get a chance.  Once you've selected a PR, you have the option of checking out as detached HEAD, checking out as a local branch, viewing the PR's diff, or opening in a browser.

**Comment by stevage** [comment]
Wow, that's really excellent. The options to checkout as detached HEAD or new branch are great. A  couple of little comments:

- The menu item "Github: Pull request" may be confusing to people who think they can open a pull request here. Maybe "Github: View pull request" (or checkout, or review...)
- Because of the way the "Git: merge" command is set up (merging another branch into current), it's not possible (I think?) to merge the PR branch into master. It would be ideal if there was an easy to find next command. Perhaps "Github: Merge pull request", which then asks what you want to merge it into, then offers to delete the local (and remote?) branch.

**Comment by stevage** [comment]
One quibble: I'm seeing "Created by TerriaJS" for most of the PRs. TerriaJS is the organisation (and also the repo name). Is it possible to get the actual contributor name?

![screenshot 2015-12-07 22 07 11](https://cloud.githubusercontent.com/assets/678344/11625403/e4df2b1c-9d2e-11e5-8b54-845c430387b3.png)

And of course, friendly times ("3 days ago") would be much nicer than raw timestamps.

**Comment by stevage** [comment]
Oh, you can fix the first bit by changing:

`user=pr["head"]["user"]["login"],`

to 

`user=pr["user"]["login"],`

**Comment by stevage** [comment]
Ok, I opened a PR. I don't do much Python these days, and I don't think I've ever written Python 3. Eep. Feel free to criticise :)

**Comment by divmain** [comment]
Thanks for catching `user=pr["user"]["login"]`, @stevage!  Your change is much better, and what I intended :)

There is an existing friendly/fuzzy date module (used for blame and log, etc), so I pulled down your changes and utilized that instead.  I made some cosmetic changes to the entries to make them easier to read.  And also added an option to create a local branch that corresponds to the PR source, but without checking it out.  This could be followed up by `git: merge` if you're so inclined, and opens up potential for other workflows also.
